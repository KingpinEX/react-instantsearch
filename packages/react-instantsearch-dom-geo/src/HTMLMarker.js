import { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import createHTMLMarker from './elements/createHTMLMarker';
import { registerEvents, createListenersPropType } from './utils';
import { GeolocHitPropType } from './propTypes';
import { GOOGLE_MAPS_CONTEXT } from './GoogleMaps';

const eventTypes = {
  onClick: 'click',
  onDoubleClick: 'dblclick',
  onMouseDown: 'mousedown',
  onMouseEnter: 'mouseenter',
  onMouseLeave: 'mouseleave',
  onMouseMove: 'mousemove',
  onMouseOut: 'mouseout',
  onMouseOver: 'mouseover',
  onMouseUp: 'mouseup',
};

class HTMLMarker extends Component {
  static propTypes = {
    ...createListenersPropType(eventTypes),
    hit: GeolocHitPropType.isRequired,
    children: PropTypes.node.isRequired,
    options: PropTypes.object,
  };

  static contextTypes = {
    [GOOGLE_MAPS_CONTEXT]: PropTypes.shape({
      google: PropTypes.object,
      instance: PropTypes.object,
    }),
  };

  static defaultProps = {
    options: {},
  };

  static isReact16() {
    return typeof ReactDOM.createPortal === 'function';
  }

  state = {
    marker: null,
  };

  componentDidMount() {
    const { hit, options } = this.props;
    const { google, instance } = this.context[GOOGLE_MAPS_CONTEXT];
    // Not the best way to create the reference of the HTMLMarker
    // but since the Google object is required didn't find another
    // solution. Ideas?
    const Marker = createHTMLMarker(google);

    const marker = new Marker({
      map: instance,
      position: hit._geoloc,
      ...options,
    });

    this.removeListeners = registerEvents(eventTypes, this.props, marker);

    this.setState(() => ({
      marker,
    }));
  }

  componentDidUpdate() {
    const { children } = this.props;
    const { marker } = this.state;

    this.removeListeners();

    this.removeListeners = registerEvents(eventTypes, this.props, marker);

    if (!HTMLMarker.isReact16()) {
      ReactDOM.unstable_renderSubtreeIntoContainer(
        this,
        children,
        marker.element
      );
    }
  }

  componentWillUnmount() {
    const { marker } = this.state;

    if (!HTMLMarker.isReact16()) {
      ReactDOM.unmountComponentAtNode(marker.element);
    }

    marker.setMap(null);
  }

  render() {
    const { children } = this.props;
    const { marker } = this.state;

    if (!marker || !HTMLMarker.isReact16()) {
      return null;
    }

    return ReactDOM.createPortal(children, marker.element);
  }
}

export default HTMLMarker;
