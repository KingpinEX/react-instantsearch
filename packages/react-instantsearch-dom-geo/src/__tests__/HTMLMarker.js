import React from 'react';
import ReactDOM from 'react-dom';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {
  createFakeGoogleReference,
  createFakeMapInstance,
  createFakeHTMLMarkerInstance,
} from '../../test/mockGoogleMaps';
import createHTMLMarker from '../elements/createHTMLMarker';
import { GOOGLE_MAPS_CONTEXT } from '../GoogleMaps';
import HTMLMarker from '../HTMLMarker';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('../elements/createHTMLMarker', () => jest.fn());

describe('HTMLMarker', () => {
  const defaultProps = {
    hit: {
      _geoloc: {
        lat: 10,
        lng: 12,
      },
    },
  };

  describe('creation', () => {
    it('expect to create the marker on didMount with default options', () => {
      const marker = createFakeHTMLMarkerInstance();
      const factory = jest.fn(() => marker);
      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      createHTMLMarker.mockImplementationOnce(() => factory);

      const props = {
        ...defaultProps,
      };

      const wrapper = shallow(
        <HTMLMarker {...props}>
          <span>This is the children.</span>
        </HTMLMarker>,
        {
          context: {
            [GOOGLE_MAPS_CONTEXT]: {
              instance: mapInstance,
              google,
            },
          },
        }
      );

      expect(createHTMLMarker).toHaveBeenCalledWith(google);
      expect(wrapper.state('marker')).toBe(marker);

      expect(factory).toHaveBeenCalledTimes(1);
      expect(factory).toHaveBeenCalledWith(
        expect.objectContaining({
          map: mapInstance,
          position: {
            lat: 10,
            lng: 12,
          },
        })
      );
    });

    it('expect to create the marker on didMount with given options', () => {
      const marker = createFakeHTMLMarkerInstance();
      const factory = jest.fn(() => marker);
      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      createHTMLMarker.mockImplementationOnce(() => factory);

      const props = {
        ...defaultProps,
        options: {
          className: 'my-marker',
          anchor: {
            x: 10,
            y: 10,
          },
        },
      };

      shallow(
        <HTMLMarker {...props}>
          <span>This is the children.</span>
        </HTMLMarker>,
        {
          context: {
            [GOOGLE_MAPS_CONTEXT]: {
              instance: mapInstance,
              google,
            },
          },
        }
      );

      expect(factory).toHaveBeenCalledWith(
        expect.objectContaining({
          className: 'my-marker',
          anchor: {
            x: 10,
            y: 10,
          },
        })
      );
    });
  });

  describe('delete', () => {
    it('expect to remove the Marker on willUnmount', () => {
      const marker = createFakeHTMLMarkerInstance();
      const factory = jest.fn(() => marker);
      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      createHTMLMarker.mockImplementationOnce(() => factory);

      const props = {
        ...defaultProps,
      };

      const wrapper = shallow(
        <HTMLMarker {...props}>
          <span>This is the children.</span>
        </HTMLMarker>,
        {
          context: {
            [GOOGLE_MAPS_CONTEXT]: {
              instance: mapInstance,
              google,
            },
          },
        }
      );

      wrapper.unmount();

      expect(marker.setMap).toHaveBeenCalledTimes(1);
      expect(marker.setMap).toHaveBeenCalledWith(null);
    });

    it("expect to remove the Marker on willUnmount only when it's created", () => {
      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      const props = {
        ...defaultProps,
      };

      const wrapper = shallow(
        <HTMLMarker {...props}>
          <span>This is the children.</span>
        </HTMLMarker>,
        {
          disableLifecycleMethods: true,
          context: {
            [GOOGLE_MAPS_CONTEXT]: {
              instance: mapInstance,
              google,
            },
          },
        }
      );

      wrapper.unmount();

      expect(() => wrapper.unmount()).not.toThrow();
    });
  });

  describe('with portal', () => {
    it('expect to render correctly', () => {
      const unstableRenderSubtreeIntoContainer = jest.spyOn(
        ReactDOM,
        'unstable_renderSubtreeIntoContainer'
      );

      const marker = createFakeHTMLMarkerInstance();
      const factory = jest.fn(() => marker);
      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      createHTMLMarker.mockImplementationOnce(() => factory);

      const props = {
        ...defaultProps,
      };

      // Use `mount` instead of `shallow` to trigger the render
      // of createPortal otherwise the Snapshot is empty
      const wrapper = mount(
        <HTMLMarker {...props}>
          <span>This is the children.</span>
        </HTMLMarker>,
        {
          context: {
            [GOOGLE_MAPS_CONTEXT]: {
              instance: mapInstance,
              google,
            },
          },
        }
      );

      expect(wrapper).toMatchSnapshot();

      expect(unstableRenderSubtreeIntoContainer).not.toHaveBeenCalled();

      unstableRenderSubtreeIntoContainer.mockReset();
      unstableRenderSubtreeIntoContainer.mockRestore();
    });

    it('expect to render correctly without a marker', () => {
      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      const props = {
        ...defaultProps,
      };

      const wrapper = shallow(
        <HTMLMarker {...props}>
          <span>This is the children.</span>
        </HTMLMarker>,
        {
          disableLifecycleMethods: true,
          context: {
            [GOOGLE_MAPS_CONTEXT]: {
              instance: mapInstance,
              google,
            },
          },
        }
      );

      expect(wrapper.type()).toBe(null);
    });

    it('expect to not render on didUpdate', () => {
      const unstableRenderSubtreeIntoContainer = jest.spyOn(
        ReactDOM,
        'unstable_renderSubtreeIntoContainer'
      );

      const marker = createFakeHTMLMarkerInstance();
      const factory = jest.fn(() => marker);
      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      createHTMLMarker.mockImplementationOnce(() => factory);

      const props = {
        ...defaultProps,
      };

      // Use `mount` instead of `shallow` to trigger didUpdate
      const wrapper = mount(
        <HTMLMarker {...props}>
          <span>This is the children.</span>
        </HTMLMarker>,
        {
          context: {
            [GOOGLE_MAPS_CONTEXT]: {
              instance: mapInstance,
              google,
            },
          },
        }
      );

      expect(unstableRenderSubtreeIntoContainer).not.toHaveBeenCalled();

      wrapper.instance().componentDidUpdate();

      expect(unstableRenderSubtreeIntoContainer).not.toHaveBeenCalled();

      unstableRenderSubtreeIntoContainer.mockReset();
      unstableRenderSubtreeIntoContainer.mockRestore();
    });

    it('expect to not call unmountComponentAtNode on willUnmount', () => {
      const unmountComponentAtNode = jest.spyOn(
        ReactDOM,
        'unmountComponentAtNode'
      );

      const marker = createFakeHTMLMarkerInstance();
      const factory = jest.fn(() => marker);
      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      createHTMLMarker.mockImplementationOnce(() => factory);

      const props = {
        ...defaultProps,
      };

      const wrapper = shallow(
        <HTMLMarker {...props}>
          <span>This is the children.</span>
        </HTMLMarker>,
        {
          context: {
            [GOOGLE_MAPS_CONTEXT]: {
              instance: mapInstance,
              google,
            },
          },
        }
      );

      wrapper.unmount();

      expect(unmountComponentAtNode).not.toHaveBeenCalled();

      unmountComponentAtNode.mockReset();
      unmountComponentAtNode.mockRestore();
    });
  });

  describe('with unstable_renderSubtreeIntoContainer', () => {
    it('expect to render correctly', () => {
      const unstableRenderSubtreeIntoContainer = jest.spyOn(
        ReactDOM,
        'unstable_renderSubtreeIntoContainer'
      );

      const isReact16 = jest
        .spyOn(HTMLMarker, 'isReact16')
        .mockImplementation(() => false);

      const marker = createFakeHTMLMarkerInstance();
      const factory = jest.fn(() => marker);
      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      createHTMLMarker.mockImplementationOnce(() => factory);

      const props = {
        ...defaultProps,
      };

      const wrapper = mount(
        <HTMLMarker {...props}>
          <span>This is the children.</span>
        </HTMLMarker>,
        {
          context: {
            [GOOGLE_MAPS_CONTEXT]: {
              instance: mapInstance,
              google,
            },
          },
        }
      );

      expect(wrapper).toMatchSnapshot();

      expect(unstableRenderSubtreeIntoContainer).toHaveBeenCalledTimes(1);
      expect(unstableRenderSubtreeIntoContainer).toHaveBeenCalledWith(
        wrapper.instance(),
        <span>This is the children.</span>,
        marker.element
      );

      unstableRenderSubtreeIntoContainer.mockReset();
      unstableRenderSubtreeIntoContainer.mockRestore();

      isReact16.mockReset();
      isReact16.mockRestore();
    });

    it('expect to render correctly without a marker', () => {
      const isReact16 = jest
        .spyOn(HTMLMarker, 'isReact16')
        .mockImplementation(() => false);

      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      const props = {
        ...defaultProps,
      };

      const wrapper = shallow(
        <HTMLMarker {...props}>
          <span>This is the children.</span>
        </HTMLMarker>,
        {
          disableLifecycleMethods: true,
          context: {
            [GOOGLE_MAPS_CONTEXT]: {
              instance: mapInstance,
              google,
            },
          },
        }
      );

      expect(wrapper.type()).toBe(null);

      isReact16.mockReset();
      isReact16.mockRestore();
    });

    it('expect to render on didUpdate', () => {
      const unstableRenderSubtreeIntoContainer = jest.spyOn(
        ReactDOM,
        'unstable_renderSubtreeIntoContainer'
      );

      const isReact16 = jest
        .spyOn(HTMLMarker, 'isReact16')
        .mockImplementation(() => false);

      const marker = createFakeHTMLMarkerInstance();
      const factory = jest.fn(() => marker);
      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      createHTMLMarker.mockImplementationOnce(() => factory);

      const props = {
        ...defaultProps,
      };

      // Use `mount` instead of `shallow` to trigger didUpdate
      const wrapper = mount(
        <HTMLMarker {...props}>
          <span>This is the children.</span>
        </HTMLMarker>,
        {
          context: {
            [GOOGLE_MAPS_CONTEXT]: {
              instance: mapInstance,
              google,
            },
          },
        }
      );

      expect(unstableRenderSubtreeIntoContainer).toHaveBeenCalledTimes(1);
      expect(unstableRenderSubtreeIntoContainer).toHaveBeenCalledWith(
        wrapper.instance(),
        <span>This is the children.</span>,
        marker.element
      );

      wrapper.instance().componentDidUpdate();

      expect(unstableRenderSubtreeIntoContainer).toHaveBeenCalledTimes(2);
      expect(unstableRenderSubtreeIntoContainer).toHaveBeenCalledWith(
        wrapper.instance(),
        <span>This is the children.</span>,
        marker.element
      );

      unstableRenderSubtreeIntoContainer.mockReset();
      unstableRenderSubtreeIntoContainer.mockRestore();

      isReact16.mockReset();
      isReact16.mockRestore();
    });

    it('expect to not render on didUpdate without marker', () => {
      const unstableRenderSubtreeIntoContainer = jest.spyOn(
        ReactDOM,
        'unstable_renderSubtreeIntoContainer'
      );

      const isReact16 = jest
        .spyOn(HTMLMarker, 'isReact16')
        .mockImplementation(() => false);

      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      const props = {
        ...defaultProps,
      };

      const wrapper = shallow(
        <HTMLMarker {...props}>
          <span>This is the children.</span>
        </HTMLMarker>,
        {
          disableLifecycleMethods: true,
          context: {
            [GOOGLE_MAPS_CONTEXT]: {
              instance: mapInstance,
              google,
            },
          },
        }
      );

      expect(unstableRenderSubtreeIntoContainer).not.toHaveBeenCalled();

      wrapper.instance().componentDidUpdate();

      expect(unstableRenderSubtreeIntoContainer).not.toHaveBeenCalled();

      unstableRenderSubtreeIntoContainer.mockReset();
      unstableRenderSubtreeIntoContainer.mockRestore();

      isReact16.mockReset();
      isReact16.mockRestore();
    });

    it('expect to call unmountComponentAtNode on willUnmount', () => {
      const unmountComponentAtNode = jest.spyOn(
        ReactDOM,
        'unmountComponentAtNode'
      );

      const isReact16 = jest
        .spyOn(HTMLMarker, 'isReact16')
        .mockImplementation(() => false);

      const marker = createFakeHTMLMarkerInstance();
      const factory = jest.fn(() => marker);
      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      createHTMLMarker.mockImplementationOnce(() => factory);

      const props = {
        ...defaultProps,
      };

      const wrapper = shallow(
        <HTMLMarker {...props}>
          <span>This is the children.</span>
        </HTMLMarker>,
        {
          context: {
            [GOOGLE_MAPS_CONTEXT]: {
              instance: mapInstance,
              google,
            },
          },
        }
      );

      wrapper.unmount();

      expect(unmountComponentAtNode).toHaveBeenCalledTimes(1);
      expect(unmountComponentAtNode).toHaveBeenCalledWith(marker.element);

      unmountComponentAtNode.mockReset();
      unmountComponentAtNode.mockRestore();

      isReact16.mockReset();
      isReact16.mockRestore();
    });
  });
});
