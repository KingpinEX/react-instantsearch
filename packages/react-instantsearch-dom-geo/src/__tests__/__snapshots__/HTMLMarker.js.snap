// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTMLMarker with portal expect to render correctly 1`] = `
<HTMLMarker
  anchor={
    Object {
      "x": 0,
      "y": 0,
    }
  }
  className=""
  hit={
    Object {
      "_geoloc": Object {
        "lat": 10,
        "lng": 12,
      },
    }
  }
>
  <span>
    This is the children.
  </span>
</HTMLMarker>
`;

exports[`HTMLMarker with unstable_renderSubtreeIntoContainer expect to render correctly 1`] = `
<HTMLMarker
  anchor={
    Object {
      "x": 0,
      "y": 0,
    }
  }
  className=""
  hit={
    Object {
      "_geoloc": Object {
        "lat": 10,
        "lng": 12,
      },
    }
  }
/>
`;
